crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,275 @@
+that the type is a singleton type.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+An example of a singleton type that is not a single-value type
+would be `Literal[Ham.A]` for the following `Ham` enum:
+because the `Ham` class overrides `__eq__`, we can no longer say for sure
+that the `Ham` type can be safely narrowed using inequality as in the above example
+(but we *can* still narrow the `Ham` type using identity, as with all singleton types):
+
+```py
+from enum import Enum, auto
+
+class Ham(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+    def __eq__(self, other):
+        return True
+```
+
+## Final types
+
+A final type in Python is a [nominal type](https://docs.python.org/3/library/typing.html#nominal-vs-structural-subtyping)
+defined as the instances of a certain runtime class that it is known cannot be subclassed.
+Most final types in Python are associated with class definitions that are decorated with
+[`@typing.final`](https://docs.python.org/3/library/typing.html#typing.final).
+This is a special decorator which indicates to the type checker that the tool
+should assume the class cannot be subclassed, and that the tool should emit an error
+if it detects an attempt to do so.
+
+There are fewer ways in which final types can be subtyped when compared to most other types.
+For any two nominal types `X` and `Y`, if `X` is `@final` and `X` is not a subtype of `Y`,
+the intersection `X & Y` is equivalent to `Never`, the empty set.
+However, a final type in Python still has subtypes, just no *nominal* subtypes.
+For the following two types `X` and `Truthy`, the intersection `X & Truthy` is non-empty,
+since `Truthy` is a structural type:
+
+```py
+from typing import final, Protocol, Literal
+
+class Truthy(Protocol):
+    """Any object `x` inhabits this type if `bool(x)` always returns `True`"""
+
+    def __bool__(self) -> Literal[True]: ...
+
+@final
+class X[T: bool]:
+    def __init__(self, bool_val: T):
+        self.bool_val = bool_val
+
+    def __bool__(self) -> T:
+        return self.bool_val
+```
+
+Not all final types are decorated with `@final`. The most common example of this
+is enum classes. Any enum class that has at least one member is considered
+implicitly final, as attempting to subclass such a class will fail at runtime:
```

Comment:
Is there a missing code example following this? Or should the colon be a full stop instead?

Reply:
Oh oops, I added a code example and then deleted it again because it didn't really add anything. I'll get rid of the colon.